File: 지은/Lv.2/땅따먹기.py
# dp
def solution(land):
    for i in range(1, len(land)):
        for j in range(4):
            max_value = 0
            for k in range(4):
                if j != k:
                    max_value = max(max_value, land[i-1][k])
            land[i][j] += max_value

    return max(land[-1])


# 다른 사람의 풀이
def solution(land):

    for i in range(1, len(land)):
        for j in range(len(land[0])):
            land[i][j] = max(land[i - 1][: j] + land[i - 1]
                             [j + 1:]) + land[i][j]

    return max(land[-1])

File: 화연/Lv.2/땅따먹기.js
/**
 * https://school.programmers.co.kr/learn/courses/30/lessons/12913
 * @param {number[][]} land
 * @returns {number}
 */
function solution(land) {
  for (let i = 1; i < land.length; i++) {
    land[i][0] += Math.max(land[i - 1][1], land[i - 1][2], land[i - 1][3]);
    land[i][1] += Math.max(land[i - 1][0], land[i - 1][2], land[i - 1][3]);
    land[i][2] += Math.max(land[i - 1][0], land[i - 1][1], land[i - 1][3]);
    land[i][3] += Math.max(land[i - 1][0], land[i - 1][1], land[i - 1][2]);
  }

  return Math.max(...land[land.length - 1]);
}
